File: Striver/DP/Q1.java
package Striver.DP;

public class Q1 {
    // fibonacci number

    //recursive solution - Time complexity: O(2^n), Space complexity: O(n)
    public static int fib(int n){
        if(n==0 || n==1) return n;
        return fib(n-1) + fib(n-2);
    }

    //memoization - Time complexity: O(n), Space complexity: O(n)
    public static int fib_memo(int n, int[] dp){
        if(n==0 || n==1) return n;
        if(dp[n]!=-1) return dp[n];
        dp[n] = fib_memo(n-1, dp) + fib_memo(n-2, dp);
        return dp[n];
    }

    //tabulation - Time complexity: O(n), Space complexity: O(n)
    public static int fib_tab(int n){
        int[] dp = new int[n+1];
        dp[0] = 0;
        dp[1] = 1;
        for(int i=2; i<=n; i++){
            dp[i] = dp[i-1] + dp[i-2];
        }
        return dp[n];
    }

    //space optimization - Time complexity: O(n), Space complexity: O(1)
    public static int fib_sp(int n){
        int a = 0, b = 1, c = 0;
        for(int i=2; i<=n; i++){
            c = a + b;
            a = b;
            b = c;
        }
        return b;
    }

    public static void main(String[] args) {
        int n = 10;
        int[] dp = new int[n+1];
        for(int i=0; i<=n; i++) dp[i] = -1;
        System.out.println(fib(n));
        System.out.println(fib_memo(n, dp));
        System.out.println(fib_tab(n));
        System.out.println(fib_sp(n));
    }
}

File: Striver/DP/Q1.py
# nth fibonacci number

# recursion - Time: O(2^n) Space: O(n)
def fib(n):
    if n == 0 or n == 1:
        return n
    return fib(n-1) + fib(n-2)

# memoization - Time: O(n) Space: O(n)
def fib(n,dp):
    if n == 0 or n == 1:
        return n
    if dp[n] != -1:
        return dp[n]
    dp[n] = fib(n-1,dp) + fib(n-2,dp)
    return dp[n]

# tabulation - Time: O(n) Space: O(n)
def fib(n):
    dp = [0,1]
    for i in range(2,n+1):
        dp.append(dp[i-1] + dp[i-2])
    return dp[n]

# space optimization - Time: O(n) Space: O(1)
def fib(n):
    a = 0
    b = 1
    for i in range(2,n+1):
        c = a + b
        a = b
        b = c
    return b


n = int(input())
dp = [-1]*(n+1)
print(fib(n,dp))

